Zend/tests/constants/final_constants/final_const3.phpt

```diff
@@ -0,0 +1,13 @@
+--TEST--
+Private class constants cannot be final
+--FILE--
+<?php
+
+class Foo
+{
+    private final const A = "foo";
+}
+
+?>
+--EXPECTF--
+Fatal error: Private constant Foo::A cannot be final as it is never overridden in %s on line %d
```

Comment:
Not really a major issue since it's impossible for existing code to use `private final const` so there's no bc break.

It seems allowed to override private constants with public constants and static:: will use the public constants, e.g. in php 7.4

Also, the fact that `private final` is forbidden isn't documented in the rfc right now? https://wiki.php.net/rfc/inheritance_private_methods

```
php > class A { public const A = 'const A::A'; public static function dump() { var_dump(static::A); } }
php > class B extends A { public const A = 'const B::A'; }
php > B::dump();
string(10) "const B::A"


php > class X { private function foo() { echo "in X::foo\n";} public function dump() { $this->foo(); } }
php > class Y extends X { public function foo() { echo "In Y::foo\n"; } }
php > (new Y())->dump();
in X::foo
```

-------

I guess updating the issue message may be better, e.g. `as it would prevent subclasses from accessing a constant of that name` but I don't know why that would be a compelling reason to forbid `private final` - it'd also be a compile error if those subclasses tried to declare the constant to override `protected final`

Somewhat reminds me of https://externals.io/message/110540#110576

Reply:
First of all, thank you for the review!

> Also, the fact that private final is forbidden isn't documented in the rfc right now?

Yeah, this is quite unfortunate... I realized that it's not documented just after starting the vote. :(

> I don't know why that would be a compelling reason to forbid private final - it'd also be a compile error if those subclasses tried to declare the constant to override protected final

The reason why `final private` is forbidden is to be consistent with https://wiki.php.net/rfc/inheritance_private_methods indeed. So In my opinion (which might be wrong of course) if private methods (- constructors) cannot be final then the same should apply for constants. Unfortunately, I'm not 100% sure what exact issue is that you see, so could you please clarify it?